# Contents:

#### Library import and df setup
#### Addressing missing values in df prods
#### Addressing Duplicates in df prods
#### Exporting as CSV file
#### Check for mixed-type data in df ords
#### Run a check for duplicate values in  df ords
#### Exporting as CSV file

## Library import and df setup

In [1]:
# To import libraries
import pandas as pd
import numpy as np
import os

In [2]:
# To establish path to Instacart project folder
path=r'C:\Users\Shelb\OneDrive\Documents\CF Coursework\Python Fundamentals for Data Analysts\10-2022 Instacart Basket Analysis'

In [3]:
# Establish products.csv file as a dataframe object
df_prods=pd.read_csv(os.path.join(path,'Data','Original Data','Products.csv'),index_col=False)

In [20]:
# Establish orders_wrangled.csv file as dataframe object
df_ords=pd.read_csv(os.path.join(path,'Data','Prepared Data','orders_wrangled.csv'),index_col=False)

## Addressing missing values

In [4]:
# To find missing values in df_prods
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [6]:
# To create a subset of null values in the df_prods dataframe
df_nan=df_prods[df_prods['product_name'].isnull()==True]
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [7]:
# To find the number of rows and columns of df_prods
df_prods.shape

(49693, 5)

In [9]:
# To create a products dataframe with no null values
df_prods_clean=df_prods[df_prods['product_name'].isnull()==False]

In [10]:
# To ensure that df_prods_clean has removed the 16 null rows
df_prods_clean.shape

(49677, 5)

## Addressing Duplicates

In [11]:
# To look for full duplicates in df_prods_clean
df_dups=df_prods_clean[df_prods_clean.duplicated()]
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [15]:
# To remove full duplicates from df_prods_clean
df_prods_clean_nodups=df_prods_clean.drop_duplicates()

In [16]:
df_prods_clean_nodups.shape

(49672, 5)

## Exporting as CSV file

In [17]:
# To export df_prods_clean_nodups to the Prepared Data folder as a csv file
df_prods_clean_nodups.to_csv(os.path.join(path,'Data','Prepared Data','products_checked.csv'))

In [18]:
# To generate descriptive statistics for the cleaned df_prods dataframe
df_prods_clean_nodups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

## The max of the price column should be investigated further. Nearly 100k is a high price for a grocery product.

## Check for mixed-type data in your df_ords dataframe

In [22]:
# To check for mixed-type data in df_ords
for col in df_ords.columns.tolist():
    weird=(df_ords[[col]].applymap(type)!=df_ords[[col]].iloc[0].apply(type)).any(axis=1)
    if len(df_ords[weird])>0:
        print(col)

## There are no columns with mixed-type data in the df_ords dataframe

## Run a check for missing values in your df_ords dataframe

In [23]:
# To find missing values in the orders dataframe
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
order_day_of_week              0
hour_order_placed              0
days_since_prior_order    206209
dtype: int64

## There are 206209 observations with missing data in the days_since_prior_order column. This could be because the observations refer to new users who have never used Instacart before.

In [24]:
# To create a new dataframe and add a boolean type column to identify new customers
df_ords_clean=df_ords
df_ords_clean['new_customer']=df_ords['days_since_prior_order'].isnull()==True
df_ords_clean

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0                 0   2539329        1             1                  2   
1                 1   2398795        1             2                  3   
2                 2    473747        1             3                  3   
3                 3   2254736        1             4                  4   
4                 4    431534        1             5                  4   
...             ...       ...      ...           ...                ...   
3421078     3421078   2266710   206209            10                  5   
3421079     3421079   1854736   206209            11                  4   
3421080     3421080    626363   206209            12                  1   
3421081     3421081   2977660   206209            13                  1   
3421082     3421082    272231   206209            14                  6   

         hour_order_placed  days_since_prior_order  new_customer  
0                        8                     NaN          True  
1                        7                    15.0         False  
2                       12                    21.0         False  
3                        7                    29.0         False  
4                       15                    28.0         False  
...                    ...                     ...           ...  
3421078                 18                    29.0         False  
3421079                 10                    30.0         False  
3421080                 12                    18.0         False  
3421081                 12                     7.0         False  
3421082                 14                    30.0         False  

[3421083 rows x 8 columns]

## I used a new variable to flag missing values in the days_since_prior_order column because the missing values in this column represent important information that can be used in an analysis

## Run a check for duplicate values in your df_ords data

In [25]:
# To look for full duplicates in df_ords_clean
df_dups=df_ords_clean[df_ords_clean.duplicated()]
df_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, order_day_of_week, hour_order_placed, days_since_prior_order, new_customer]
Index: []

## There are no duplicate values in the df_ords dataframe

## Exporting as CSV file

In [26]:
df_ords_clean.to_csv(os.path.join(path,'Data','Prepared Data','orders_checked.csv'))